src/lxc/lxccontainer.c

```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	template = malloc(len);
+	if (!template) {
+		SYSERROR("Not enough memory");
+		goto out_err_free;
+	}
+	ret = snprintf(template, len, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
+	if (ret < 0 || ret >= len) {
+		SYSERROR("Error writing shmounts tempdir name");
+		goto out_err_free;
+	}
+
+	/* Create a temporary dir / file
+	 * under the shared mountpoint
+	 */
+	ret = stat(source, &sb);
+	if (ret < 0) {
+		SYSERROR("Could not read source information");
+		goto out_err_free;
+	}
+
+	if (S_ISREG(sb.st_mode)) {
+		fd = mkstemp(template);
+		if (fd < 0) {
+			SYSERROR("Error creating shmounts temporary file");
+			goto out_err_free;
+		}
+	} else if (S_ISDIR(sb.st_mode)) {
+		template = mkdtemp(template);
+		if (template == NULL) {
+			SYSERROR("Error creating shmounts tempdir");
+			goto out_err_free;
+		}
+	} else {
+		ERROR("Unexpected type of shared mountpoint");
+		goto out_err_free;
+	}
+
+	/* Do the fork */
+	pid = fork();
+	if (pid < 0) {
+		SYSERROR("Could not fork");
+		goto out_err_free;
+	}
+
+	if (pid == 0) {
+		int exit_status = EXIT_FAILURE;
+
+		/* Do the mount */
+		ret = mount(source, template, filesystemtype, mountflags, data);
+		if (ret < 0) {
+			SYSERROR("Failed to mount \"%s\" onto \"%s\"", source, template);
+			goto out_child_err_temp;
+		}
+
+		init_pid = do_lxcapi_init_pid(c);
+		if (init_pid < 0)
+			goto out_child_err_temp;
+
+		/* Enter the container namespaces */
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			/* if fully unprivileged */
+			ret = setns(init_pid,CLONE_NEWUSER);
+			if (ret < 0) {
+				SYSERROR("Failed to join container's userns");
+				goto out_child_err_temp;
+			}
+		}
+		ret = setns(init_pid, CLONE_NEWNS);
+		if (ret < 0) {
+			SYSERROR("Failed to join container's mountns");
+			goto out_child_err_temp;
+		}
+
+		suff = strrchr(template, '/');
+		if (!suff)
+			goto out_child_err_temp;
+
+		len = strlen(c->lxc_conf->lxc_shmount.path_cont) + (sizeof("/lxcmountXXXXXX") - 1);
+		path = malloc(len);
+		if (!path) {
+			SYSERROR("Not enough memory");
+			goto out_child_err_temp;
+		}
+		ret = snprintf(path, len, "%s%s", c->lxc_conf->lxc_shmount.path_cont, suff);
+		if (ret < 0 || ret >= len) {
+			SYSERROR("Error writing container mountpoint name");
+			goto out_child_err_path;
+		}
+
+		ret = mkdir_p(target, 0700);
+		if (ret < 0) {
+			ERROR("Failed to create container temp mountpoint");
+			goto out_child_err_path;
+		}
+
+		ret = mount(path, target, NULL, MS_REC | MS_MOVE, NULL);
+		if (ret < 0) {
+			SYSERROR("Failed to move the mount");
+			goto out_child_err_path;
+		}
+
+		exit_status = EXIT_SUCCESS;
+
+out_child_err_path:
+		free(path);
+
+out_child_err_temp:
+		free(template);
+
+		_exit(exit_status);
+	}
+
+	ret = waitpid(pid, &status, 0);
+	if (ret < 0) {
+		SYSERROR("Wait for the chile with pid %ld failed", (long) pid);
+		goto out_err_free;
+	}
+	if (WIFEXITED(status) && WEXITSTATUS(status) != EXIT_SUCCESS)
+		goto out_err_unmount;
```

Comment:
You should use our helper [wait_for_pid](https://github.com/lxc/lxc/blob/83ffaa1d39096ee2897c6f35d523a819698d1daa/src/lxc/utils.c#L278) which will do all of that for you.